File: querier/internal/query/manifest_cache.go
package query

import (
	"context"
	"database/sql"
	"fmt"
	"os"
	"sync"
)

// cachedManifestRow is one shards-table row, holding just enough to satisfy
// both loadSourceJars and loadManifestSources without reopening and rescanning
// the manifest on every query.
type cachedManifestRow struct {
	shardID    string
	coordinate string // "" when the jar has no recorded Maven coordinate
	filename   string
	source     string // COALESCE(NULLIF(coordinate, ''), filename)
}

// The manifest changes only when an indexer/prefetch run rewrites it, so its
// shard rows are cached process-wide and keyed by path+mtime. A long-lived MCP
// server reuses the parsed rows across tool calls; a concurrent index run that
// rewrites the manifest bumps the mtime and is picked up on the next call. We
// cache the parsed data, never a live mutable *sql.DB handle.
var (
	manifestCacheMu    sync.Mutex
	manifestCachePath  string
	manifestCacheMtime int64
	manifestCacheRows  []cachedManifestRow
)

// loadManifestRows returns the manifest's shard rows, served from the process
// cache when the file's mtime is unchanged. Returns (nil, nil) when there is no
// manifest path. A stat failure disables the cache for the call but still reads
// the manifest, so correctness never depends on the cache.
func loadManifestRows(ctx context.Context, manifestPath string) ([]cachedManifestRow, error) {
	if manifestPath == "" {
		return nil, nil
	}
	info, statErr := os.Stat(manifestPath)
	if statErr == nil {
		mtime := info.ModTime().UnixNano()
		manifestCacheMu.Lock()
		if manifestCachePath == manifestPath &&
			manifestCacheMtime == mtime && manifestCacheRows != nil {
			rows := manifestCacheRows
			manifestCacheMu.Unlock()
			return rows, nil
		}
		manifestCacheMu.Unlock()
	}

	rows, err := scanManifestRows(ctx, manifestPath)
	if err != nil {
		return nil, err
	}
	if statErr == nil {
		manifestCacheMu.Lock()
		manifestCachePath = manifestPath
		manifestCacheMtime = info.ModTime().UnixNano()
		manifestCacheRows = rows
		manifestCacheMu.Unlock()
	}
	return rows, nil
}

func scanManifestRows(ctx context.Context, manifestPath string) ([]cachedManifestRow, error) {
	manifest, err := openReadOnly(manifestPath, false)
	if err != nil {
		return nil, fmt.Errorf("open manifest database: %w", err)
	}
	defer manifest.Close()
	rows, err := manifest.QueryContext(
		ctx,
		`SELECT shard_id, jar_coordinate, jar_filename,
		        COALESCE(NULLIF(jar_coordinate, ''), jar_filename)
		 FROM shards`,
	)
	if err != nil {
		return nil, fmt.Errorf("query manifest sources: %w", err)
	}
	defer rows.Close()
	result := []cachedManifestRow{}
	for rows.Next() {
		var row cachedManifestRow
		var coordinate sql.NullString
		if err := rows.Scan(&row.shardID, &coordinate, &row.filename, &row.source); err != nil {
			return nil, fmt.Errorf("scan manifest source: %w", err)
		}
		if coordinate.Valid {
			row.coordinate = coordinate.String
		}
		result = append(result, row)
	}
	return result, rows.Err()
}

File: core/src/main/kotlin/dev/research4jar/query/ManifestCache.kt
package dev.research4jar.query

import java.nio.file.Files
import java.nio.file.Paths
import java.sql.Connection

data class CachedManifestRow(
    val shardId: String,
    /** "" when the jar has no recorded Maven coordinate. */
    val coordinate: String,
    val filename: String,
    /** COALESCE(NULLIF(coordinate, ''), filename). */
    val source: String,
)

/**
 * Process-wide cache of the manifest's shards table keyed by path+mtime.
 * The manifest changes only when an indexer/prefetch run rewrites it; a
 * long-lived MCP server or daemon reuses the parsed rows across calls, and a
 * rewrite bumps the mtime and is picked up on the next call. Parsed data
 * only — never a live connection.
 */
object ManifestCache {
    private val lock = Any()
    private var cachedPath: String? = null
    private var cachedMtime: Long = 0
    private var cachedRows: List<CachedManifestRow>? = null

    fun loadRows(manifestPath: String): List<CachedManifestRow> {
        if (manifestPath.isEmpty()) return emptyList()
        val mtime = try {
            Files.getLastModifiedTime(Paths.get(manifestPath)).toMillis()
        } catch (_: java.io.IOException) {
            null
        }
        if (mtime != null) {
            synchronized(lock) {
                val rows = cachedRows
                if (cachedPath == manifestPath && cachedMtime == mtime && rows != null) {
                    return rows
                }
            }
        }
        val rows = scanRows(manifestPath)
        if (mtime != null) {
            synchronized(lock) {
                cachedPath = manifestPath
                cachedMtime = mtime
                cachedRows = rows
            }
        }
        return rows
    }

    /**
     * Resolve both canonical shard ids and a v7 session's compact numeric
     * keys. Including the canonical ids keeps hand-written/legacy test
     * sessions compatible; production result rows use the numeric entries.
     * Missing manifest rows still resolve to the canonical shard id rather
     * than leaking the internal integer into user-facing JSON.
     */
    fun loadSourceJars(
        session: Connection,
        manifestPath: String,
        shardKeys: Collection<String>,
    ): Map<String, String> =
        mapSessionRows(session, loadRows(manifestPath), shardKeys).mapValues { (_, row) ->
            row.source.ifEmpty { row.shardId }
        }

    internal fun mapSessionRows(
        session: Connection,
        manifestRows: List<CachedManifestRow>,
        shardKeys: Collection<String>? = null,
    ): Map<String, CachedManifestRow> {
        val byCanonicalId = manifestRows.associateBy(CachedManifestRow::shardId)
        val requested = shardKeys?.distinct()
        val mapped = LinkedHashMap<String, CachedManifestRow>(
            if (requested == null) byCanonicalId.size * 2 else requested.size,
        )
        if (requested == null) {
            mapped.putAll(byCanonicalId)
        } else {
            requested.forEach { key -> byCanonicalId[key]?.let { mapped[key] = it } }
        }
        try {
            val numericKeys = requested?.mapNotNull { it.toLongOrNull() }
            if (numericKeys == null) {
                session.query(
                    "SELECT shard_key, shard_id FROM session_shards",
                    emptyList(),
                ) { rows ->
                    while (rows.next()) putSessionRow(rows, byCanonicalId, mapped)
                }
            } else {
                // Public pages cap at 1000 rows. Chunking keeps well below
                // SQLite builds with a 999-variable limit while resolving only
                // the page's keys, not every shard in a 10k-jar session.
                numericKeys.distinct().chunked(400).forEach { chunk ->
                    if (chunk.isEmpty()) return@forEach
                    session.query(
                        "SELECT shard_key, shard_id FROM session_shards WHERE shard_key IN (" +
                            List(chunk.size) { "?" }.joinToString(",") + ")",
                        chunk,
                    ) { rows ->
                        while (rows.next()) putSessionRow(rows, byCanonicalId, mapped)
                    }
                }
            }
        } catch (_: java.sql.SQLException) {
            // A hand-written/pre-v7 session uses canonical text ids directly.
        }
        return mapped
    }

    private fun putSessionRow(
        rows: java.sql.ResultSet,
        byCanonicalId: Map<String, CachedManifestRow>,
        mapped: MutableMap<String, CachedManifestRow>,
    ) {
        val key = rows.getLong(1).toString()
        val shardId = rows.getString(2)
        mapped[key] = byCanonicalId[shardId] ?: CachedManifestRow(
            shardId = shardId,
            coordinate = "",
            filename = "",
            source = "",
        )
    }

    private fun scanRows(manifestPath: String): List<CachedManifestRow> =
        Db.openReadOnly(manifestPath, immutable = false).use { manifest ->
            manifest.query(
                """
                SELECT shard_id, jar_coordinate, jar_filename,
                       COALESCE(NULLIF(jar_coordinate, ''), jar_filename)
                FROM shards
                """.trimIndent(),
                emptyList(),
            ) { rows ->
                rows.mapRows {
                    CachedManifestRow(
                        shardId = it.getString(1),
                        coordinate = it.getString(2) ?: "",
                        filename = it.getString(3),
                        source = it.getString(4),
                    )
                }
            }
        }
}
